pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageIdAdvUtils.java

```diff
@@ -51,12 +51,14 @@ static boolean acknowledge(MessageIdAdv msgId, boolean individual) {
             return false;
         }
         int batchIndex = msgId.getBatchIndex();
-        if (individual) {
-            ackSet.clear(batchIndex);
-        } else {
-            ackSet.clear(0, batchIndex + 1);
+        synchronized (ackSet) {
+            if (individual) {
+                ackSet.clear(batchIndex);
+            } else {
+                ackSet.clear(0, batchIndex + 1);
+            }
+            return ackSet.isEmpty();
```

Comment:
My concern is that this isn't sufficient to address the thread safety issues.

The previous code was using synchronized for more accesses of BitSet.
https://github.com/apache/pulsar/blob/11cfaea0deefc3c5d32f6504d266f84d12f89641/pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageAcker.java#L59-L77

There's org.apache.pulsar.common.util.collections.ConcurrentBitSet in the code base, however that doesn't seem to cover the clear method at all.

Reply:
for example ConcurrentBitSet.clear is called here: https://github.com/apache/pulsar/blob/35bb021cdbd9cc9f4f801bd14f95d035a76a0043/pulsar-client/src/main/java/org/apache/pulsar/client/impl/PersistentAcknowledgmentsGroupingTracker.java#L335

Perhaps there's something else that makes this correct?

Reply:
Yes, this fix primarily ensures that individual acknowledgments (without an ackset) do not get lost.

Reply:
The root cause is that `MessageIdAdv#getAckSet` could expose the reference of `ackSet` to other threads that might modify the `ackSet` without synchronization. So it requires the caller of `getAckSet()` to apply explicit synchronization on **all** accesses to the same `BitSet` instance.

Currently, `getAckSet` could also be called in
- `PersistentAcknowledgmentsGroupingTracker#doIndividualBatchAckAsync`
- `PersistentAcknowledgmentsGroupingTracker#doImmediateBatchIndexAck`

So you have to protect these accesses as well.

```java
// doImmediateBatchIndexAck
            synchronized (msgId.getAckSet()) {
                bitSet = BitSetRecyclable.valueOf(msgId.getAckSet().toLongArray());
            }
```

```java
// doIndividualBatchAckAsync
                    final BitSet ackSet = msgId.getAckSet();
                    final ConcurrentBitSetRecyclable value;
                    if (ackSet == null) {
                        value = ConcurrentBitSetRecyclable.create();
                        value.set(0, msgId.getBatchSize());
                    } else {
                        // Note: IDE might indicates Synchronization on local variable, but it's reasonable here
                        //  because ackSet references the `BatchMessageIdImpl#ackSet` object
                        synchronized (ackSet) {
                            if (ackSet.isEmpty()) {
                                value = ConcurrentBitSetRecyclable.create();
                                value.set(0, msgId.getBatchSize());
                            } else {
                                value = ConcurrentBitSetRecyclable.create(ackSet);
                            }
                        }
                    }
```

----

BTW, we might need to note `getAckSet` should return null or a **reference** to the `BitSet` object. Otherwise, synchronization on the `BitSet` object might not make sense.

Reply:
If we use thread-safe Bitset at the beginning (receive batch message from broker), this issue won't be a problem.
https://github.com/apache/pulsar/blob/d5de7664394e5947cb7435bdaa6d161d0ca8e27c/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java#L1643-L1643

And track for `pendingIndividualBatchIndexAcks` could also be easier, WDYT? cc @lhotari @BewareMyPower 
https://github.com/apache/pulsar/blob/d5de7664394e5947cb7435bdaa6d161d0ca8e27c/pulsar-client/src/main/java/org/apache/pulsar/client/impl/PersistentAcknowledgmentsGroupingTracker.java#L322-L336

Reply:
> If we use thread-safe Bitset at the beginning (receive batch message from broker), this issue won't be a problem.

It might be right. The key point is that you should avoid composite method calls. For example, the current implementation of `MessageIdAdvUtils#acknowledge` is:

```java
        if (individual) {
            ackSet.clear(batchIndex);
        } else {
            ackSet.clear(0, batchIndex + 1);
        }
        return ackSet.isEmpty();
```

We must make the composite calls of `clear` and `isEmpty` atomic. Otherwise, if the state of `ackSet` is modified by another thread after `clear`, the `isEmpty()` might return an unexpected result.

Reply:
> We must make the composite calls of clear and isEmpty atomic. Otherwise, if the state of ackSet is modified by another thread after clear, the isEmpty() might return an unexpected result.

Oh, you are right. In this case, it's possible cause duplicate ack on the same individual message.

Reply:
> Yes, this fix can only ensures that individual acks (without an ackset) do not get lost.

@Shawyeok ok, in that case, I think we need a separate fix for concurrency bugs in PersistentAcknowledgmentsGroupingTracker . I think that it also impacts individual acks.

Reply:
I've added another synchronize blocks, feel free to take a look, thanks. @BewareMyPower @lhotari 

Unit tests should also cover the corresponding changes. I will add them subsequently.

Reply:
Regarding the issues in ConcurrentBitSet that impact PersistentAcknowledgmentsGroupingTracker, I created #22361 to fix that separately.